### PR TITLE
feat(errortracking): ignoredExceptionTypes config to skip duplicate exception captures

### LINKED ISSUE #567: Native crash autocapture double-reports fatal React Native JS errors (JavascriptException)
### Problem

When the React Native SDK enables native crash autocapture (`@posthog/react-native-plugin`, `errorTracking.autocapture.nativeCrashes`, see PostHog/posthog-js#3677), a **fatal JS error can be captured twice**:

1. The JS layer captures it via `uncaughtExceptions` autocapture (with the JS stack trace), and
2. React Native rethrows fatal JS errors on the native side as a `com.facebook.react.common.JavascriptException` on the main thread. posthog-android's uncaught-exception handler then captures it as a separate native crash (with a native stack trace).

The result is two `$exception` events for one logical error.

### Suggested fix

posthog-android should expose a way to ignore exceptions that originate from React Native's fatal-JS path, and/or skip reporting throwables of type `com.facebook.react.common.JavascriptException`, so the JS-captured event is the single source of truth.

For reference, sentry-react-native does exactly this on the native side:
```java
// React native internally throws a JavascriptException.
// we want to ignore it on the native side to avoid sending it twice.
options.addIgnoredExceptionForType(JavascriptException.class);
```
(https://github.com/getsentry/sentry-react-native/blob/f170ec355cd622d3e935478a23660ad4cb97e896/packages/core/android/src/main/java/io/sentry/react/RNSentryStart.java#L355)

### Notes

- This pairs with the iOS equivalent (sibling issue in posthog-ios: PostHog/posthog-ios#653).
- Worth verifying empirically against current RN (0.79+) release builds, since the exact rethrow path can change between RN versions.
- Discovered while reviewing PostHog/posthog-js#3677.

### PR DESCRIPTION
**Pull request by turnipdabeets**
## :bulb: Motivation and Context

Closes #567. Supersedes #569 (stale-closed; its head branch lives on a fork we can't push to, so this PR carries its commits forward instead — original authorship preserved).

When React Native apps enable native crash autocapture, a fatal JS error is captured twice: once by the JS layer, and once natively when RN rethrows it as `com.facebook.react.common.JavascriptException`. This adds `errorTrackingConfig.ignoredExceptionTypes: MutableList<Class<out Throwable>>` so the native duplicate can be suppressed (mirrors sentry-android's `addIgnoredExceptionForType`; the iOS equivalent already merged in PostHog/posthog-ios#666).

`PostHog.captureException` (autocapture + manual callers) checks the throwable and its cause chain with `Class.isInstance` — R8-safe, and the cause-chain walk covers the `PostHogThrowable` wrapper. The generic `capture("$exception", ...)` path is gated too, by matching the class names serialized in `$exception_list` (per review on #569). The uncaught-exception handler still chains either way — ignoring an exception never changes crash behavior.

On top of #569's final state, this PR:
- restores the ~199 lines of `exceptionSteps` tests that #569 accidentally deleted from `PostHogTest.kt` (the blocker that stalled it) — the test file is now purely additive
- gates the generic capture path (the remaining review ask)
- adds subclass-match and `PostHogThrowable` end-to-end tests

## :green_heart: How did you test it?

`./gradlew :posthog:test` — full suite green; `PostHogTest` runs 130 tests (7 covering `ignoredExceptionTypes`: direct match, cause-chain match, subclass match, `PostHogThrowable` wrapper, non-match passthrough, and the two generic-capture cases). `spotlessCheck` and `apiCheck` pass.

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] I updated the docs if needed.
- [x] No breaking change or entry added to the changelog.

### If releasing new changes

- [x] Ran `pnpm changeset` to generate a changeset file

## 🤖 Agent context

**Autonomy:** Human-driven (agent-assisted)

- Claude Code session directed by @turnipdabeets: reconstructed #569 on current `main`, keeping tsushanth's feature commit and the review-settled design (Class-based `isInstance` matching, check inside `captureException`) and dropping only the unrelated test deletions.
- All review threads on #569 (marandaneto, ioannisj, greptile) were re-checked against this diff; the two open items (deleted tests, generic-capture gating) are addressed here.
- An independent reviewer agent audited the diff (cause-chain edge cases, `$exception_list` name reconstruction, API compatibility); its two findings — missing subclass and `PostHogThrowable` tests — are included.
- Unverified commits were recreated signed with original authorship per repo signature rules.

### CODE REVIEW

**Comment by turnipdabeets** [comment]
@ioannisj #569 was closed due to being stale but it was really close to ready so I took those commits, resolved the last requests you had on there and reopened it here (couldn’t push to other brnach) - mind taking a look?

File: posthog/src/main/java/com/posthog/PostHog.kt [hunk 1]
```diff
@@ -718,6 +742,37 @@ public class PostHog private constructor(
...
+    ): Boolean {
+        val exceptionList = properties?.get("\$exception_list") as? List<*> ?: return false
+        val ignoredNames = ignored.map { it.name }
+        return exceptionList.any { entry ->
+            val exception = entry as? Map<*, *> ?: return@any false
+            val type = exception["type"] as? String ?: return@any false
+            val module = exception["module"] as? String
+            // module + type rejoins the runtime class name ThrowableCoercer split apart
+            val className = if (module.isNullOrEmpty()) type else "$module.$type"
+            className in ignoredNames
```
[the remark this answers is not in the document]
Comment (turnipdabeets) [reply posthog/src/main/java/com/posthog/PostHog.kt:750]: This asymmetry is inherent to the serialized path, so we're keeping exact-name matching here. For bridge events we only have class-name strings — resolving them with `Class.forName` at capture time would trigger class initialization for arbitrary event-supplied names, and still couldn't work for minified or JS-originated types. Every native throwable (autocapture and manual `captureException`) goes through the `isInstance` path and gets full subclass matching; this `$exception_list` gate is a best-effort backstop for pre-serialized events, matching the merged iOS behavior (PostHog/posthog-ios#666, exact-name). The KDoc on `ignoredExceptionTypes` documents the name-based matching for this path.

File: posthog/src/main/java/com/posthog/PostHog.kt [hunk 2]
```diff
@@ -718,6 +742,37 @@ public class PostHog private constructor(
...
+        }
+    }
+
+    private fun findIgnoredTypeInCauseChain(
+        throwable: Throwable,
+        ignored: List<Class<out Throwable>>,
+    ): Class<out Throwable>? {
+        var current: Throwable? = throwable
+        var depth = 0
+        while (current != null && depth < MAX_CAUSE_DEPTH) {
```
[the remark this answers is not in the document]
Comment (turnipdabeets) [reply posthog/src/main/java/com/posthog/PostHog.kt]: Good catch — fixed in cefe9f4: the walk now uses the same seen-set cycle detection as `ThrowableCoercer` instead of a depth cap, so the two walks always cover the same chain and an ignored type can't be serialized into `$exception_list` while escaping the `isInstance` check.

File: posthog/src/main/java/com/posthog/PostHog.kt [hunk 3]
```diff
@@ -620,6 +620,19 @@ public class PostHog private constructor(
                 groupIdentify = true
             }
 
+            // Externally-built $exception events (e.g. the Flutter/RN bridge) carry only
```
Comment (ioannisj) [inline posthog/src/main/java/com/posthog/PostHog.kt:623]: Should we run this in captureExceptionStateless as well?

File: posthog/src/main/java/com/posthog/errortracking/PostHogErrorTrackingConfig.kt [hunk 4]
```diff
@@ -37,4 +37,12 @@ public class PostHogErrorTrackingConfig
...
          */
         public val exceptionSteps: PostHogExceptionStepsConfig = PostHogExceptionStepsConfig(),
+        /**
+         * Throwable classes to drop during capture. The throwable and every cause in
+         * its chain are matched via [Class.isInstance]. Also applies to `$exception`
+         * events sent through the generic capture path (matched by class name).
+         *
+         * Defaults to empty.
+         */
+        public val ignoredExceptionTypes: MutableList<Class<out Throwable>> = mutableListOf(),
```
Comment (ioannisj) [inline posthog/src/main/java/com/posthog/errortracking/PostHogErrorTrackingConfig.kt:47]: iOS ships with a default RCTFatalException here, do we need something similar?

File: posthog/src/main/java/com/posthog/PostHog.kt (hunk 3 above)
Comment (ioannisj) [reply posthog/src/main/java/com/posthog/PostHog.kt:623]: Then again looking at posthog-ios we only capture through captureException() so there's a gap there?

**Review by ioannisj — APPROVED** [review]
LG but left a couple of comments

File: posthog/src/main/java/com/posthog/PostHog.kt (hunk 3 above)
Comment (turnipdabeets) [reply posthog/src/main/java/com/posthog/PostHog.kt:623]: should I update iOS as well?

Comment (ioannisj) [reply posthog/src/main/java/com/posthog/PostHog.kt:623]: yeah I think that we should be ignoring through capture() as well. Probably missed it in the iOS PR

Comment (ioannisj) [reply posthog/src/main/java/com/posthog/PostHog.kt:623]: For Android, there's also captureExceptionStateless exposed as a public API though

Comment (turnipdabeets) [reply posthog/src/main/java/com/posthog/PostHog.kt:623]: Both covered now:

- `captureExceptionStateless` runs the same cause-chain gate as of 6994035 — the check moved into `PostHogStateless` so both entry points share one implementation (+2 stateless tests).
- For the iOS gap: opened PostHog/posthog-ios#713. All the generic `capture(...)` overloads funnel into `captureInternal`, so the `$exception_list` check now lives there; the `captureException` copy became redundant and was removed, and the crash-report path keeps its early check to short-circuit report processing before the drop.

File: posthog/src/main/java/com/posthog/errortracking/PostHogErrorTrackingConfig.kt (hunk 4 above)
Comment (turnipdabeets) [reply posthog/src/main/java/com/posthog/errortracking/PostHogErrorTrackingConfig.kt:47]: We don't think Android can ship a safe default here: the list is `Class`-based (per the R8 discussion on #569), so a built-in default would mean referencing `JavascriptException::class` — an RN dependency in the core SDK — and a name-string default instead would break under R8 renaming, which is why name matching was rejected in the first place. Our plan is for the RN plugin to register `com.facebook.react.common.JavascriptException::class.java` when it enables native crash autocapture — same split sentry uses (core exposes the API, the RN layer supplies the default). Happy to go another way if you'd prefer.
